### ISSUE TITLE
for loop incorrectly transformed by p4c

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4c#4739]
I was writing some simple test cases of for loops to try out some aspects of this feature, and believe I have found a case where the compiler front end in some pass has an incorrect transformation, because the output of the front end, and thus also the mid end, appear functionally to be infinite loops, whereas the input program was definitely not.

I tested with two versions of p4c, the most recent of which was:

+ p4c git SHA 32e73964a43941ae10ef5ac1e0dbd3a9d7975ed3 dated 2024-Jun-21

The command I used was:
```
mkdir -p tmp
p4test --dump tmp --top4 FrontEndLast,FrontEndDump,MidEndLast loop-var-modifiable-in-body1.p4
```
The attached zip file contains the input P4 program, and the contents of the `tmp` directory created by the command above on my system.
[p4c-issue-4739.zip](https://github.com/user-attachments/files/15936674/p4c-issue-4739.zip)

**Comment by jafingerhut** [comment]
@ChrisDodd Pinging you on this one, in case you have time to look at it.

**Comment by ChrisDodd** [comment]
The problem seems to happen in DoSimplifyDefUse -- it seems that it thinks the assignments to `i` in the loop are dead (have no uses), so they get elimiinated.  Not sure why that is happening.

**Comment by ChrisDodd** [comment]
Should be fixed by #4783
